mmv1/third_party/terraform/services/apigee/resource_apigee_api.go

```diff
@@ -0,0 +1,403 @@
+
+	log.Printf("[DEBUG] sendRequestRawBodyWithTimeout Done")
+	if err != nil {
+		return fmt.Errorf("Error creating API proxy: %s", err)
+	}
+
+	// Store the ID now
+	id, err := tpgresource.ReplaceVars(d, config, "organizations/{{org_id}}/apis/{{name}}")
+	if err != nil {
+		return fmt.Errorf("Error constructing id: %s", err)
+	}
+	d.SetId(id)
+	log.Printf("[DEBUG] create d.SetId done, id = %s", id)
+
+	log.Printf("[DEBUG] Finished creating API proxy %q: %#v", d.Id(), res)
+
+	if resourceApigeeApiRead(d, meta) != nil {
+		return fmt.Errorf("Error reading API proxy at end of Create: %s", err)
+	}
+	d.Set("md5hash", localMd5Hash)
+	d.Set("detect_md5hash", localMd5Hash)
+	return nil
+}
+
+func resourceApigeeApiUpdate(d *schema.ResourceData, meta interface{}) error {
+	//For how API proxy api is implemented, just treat an update as create, when the name is same, it will create a new revision
+	return resourceApigeeApiCreate(d, meta)
+}
+
+func resourceApigeeApiRead(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*transport_tpg.Config)
+	userAgent, err := tpgresource.GenerateUserAgentString(d, config.UserAgent)
+	if err != nil {
+		return err
+	}
+
+	url, err := tpgresource.ReplaceVars(d, config, "{{ApigeeBasePath}}organizations/{{org_id}}/apis/{{name}}")
+	if err != nil {
+		return err
+	}
+	log.Printf("[DEBUG] API proxy read url is: %s", url)
+
+	billingProject := ""
+
+	// err == nil indicates that the billing_project value was found
+	if bp, err := tpgresource.GetBillingProject(d, config); err == nil {
+		billingProject = bp
+	}
+	log.Printf("[DEBUG] resourceApigeeApiRead sendRequest")
+	log.Printf("[DEBUG] resourceApigeeApiRead, url=, 	%s", url)
+	res, err := transport_tpg.SendRequest(transport_tpg.SendRequestOptions{
+		Config:    config,
+		Method:    "GET",
+		Project:   billingProject,
+		RawURL:    url,
+		UserAgent: userAgent,
+	})
+	if err != nil {
+		return transport_tpg.HandleNotFoundError(err, d, fmt.Sprintf("ApigeeApi %q", d.Id()))
+	}
+	log.Printf("[DEBUG] resourceApigeeApuRead sendRequest completed")
+	previousLastModifiedAt := getApigeeApiLastModifiedAt(d)
+	if err := d.Set("meta_data", flattenApigeeApiMetaData(res["metaData"], d, config)); err != nil {
+		return fmt.Errorf("Error reading API proxy: %s", err)
+	}
+	currentLastModifiedAt := getApigeeApiLastModifiedAt(d)
+	if err := d.Set("name", flattenApigeeApiName(res["name"], d, config)); err != nil {
+		return fmt.Errorf("Error reading API proxy: %s", err)
+	}
+	if err := d.Set("revision", flattenApigeeApiRevision(res["revision"], d, config)); err != nil {
+		return fmt.Errorf("Error reading API proxy: %s", err)
+	}
+	if err := d.Set("latest_revision_id", flattenApigeeApiLatestRevisionId(res["latestRevisionId"], d, config)); err != nil {
+		return fmt.Errorf("Error reading API proxy: %s", err)
+	}
+
+	//setting hash to suggest update
+	if previousLastModifiedAt != currentLastModifiedAt {
```

Comment:
Why do we go through the whole md5hash detection for this resource?

It seems to me that we don't read the hash from the server-side, so we're only calculating it locally. But we only calculate it locally during create, and just use it here to trigger recreate if the last modified time changes. Can we just have a customize diff on last modified at to trigger this?

Reply:
I understand that the hash is calculated and stored in the state so it can be determined whether we locally have a change from what was previously deployed using terraform from what we currently have locally. As the revision is established remotely and not locally there would not be way to understand if there was been a changed locally without the hash.